src/Runtime/Distributions/IImmutableDistribution.cs

```diff
@@ -0,0 +1,297 @@
+    [Quality(QualityBand.Experimental)]
+    public interface CanCopyWithMean<in MeanType, out TDist>
+    {
+        /// <summary>
+        /// Creates a copy of the current distribution with mean set to the provided value.
+        /// </summary>
+        /// <param name="value">The mean value</param>
+        /// <returns>The created distribution.</returns>
+        TDist WithMean(MeanType value);
+    }
+
+    /// <summary>
+    /// Whether the distribution supports creating copies with different mean and variance.
+    /// </summary>
+    /// <typeparam name="MeanType">Type of the mean value.</typeparam>
+    /// <typeparam name="VarType">Type of the variance value.</typeparam>
+    /// <typeparam name="TDist">Type of the distribution.</typeparam>
+    [Quality(QualityBand.Experimental)]
+    public interface CanCopyWithMeanAndVariance<in MeanType, in VarType, out TDist>
+    {
+        /// <summary>
+        /// Creates a copy of the current distribution with mean and variance set to the provided values.
+        /// </summary>
+        /// <param name="mean">Mean</param>
+        /// <param name="variance">Variance</param>
+        /// <returns>The created distribution.</returns>
+        /// <remarks>In cases where the given mean and variance cannot be matched exactly, the mean
+        /// should have priority.</remarks>
+        TDist WithMeanAndVariance(MeanType mean, VarType variance);
+    }
+
+    /// <summary>
+    /// Whether the distribution supports creating copies with all elements set to duplicates of the same value.
+    /// </summary>
+    /// <typeparam name="T">Type of the element.</typeparam>
+    /// <typeparam name="TDist">Type of the distribution.</typeparam>
+    [Quality(QualityBand.Experimental)]
+    public interface CanCopyWithAllElementsSetTo<in T, out TDist>
+    {
+        /// <summary>
+        /// Creates a copy of the current distribution with all elements set to duplicates of the given value.
+        /// </summary>
+        /// <param name="value">The value of all elements of the new distribution.</param>
+        /// <returns>The created distribution.</returns>
+        TDist WithAllElementsSetTo(T value);
+    }
+
+    /// <summary>
+    /// Supports computing the product of the current instance and another value.
+    /// </summary>
+    /// <typeparam name="TOther">The type of the second multiplier.</typeparam>
+    /// <typeparam name="TResult">The type of the product.</typeparam>
+    /// <remarks>Typically the type of the product would be the same as the type of the current instance.</remarks>
+    [Quality(QualityBand.Experimental)]
+    public interface CanComputeProduct<in TOther, out TResult>
+    {
+        /// <summary>
+        /// Computes the product of the current instance and <paramref name="other"/>.
+        /// </summary>
+        /// <param name="other">The second multiplier.</param>
+        /// <returns>The product.</returns>
+        TResult Product(TOther other);
```

Comment:
Method names should be verbs

Reply:
It is more common to call this `Multiply`.  See [TimeSpan.Multiply](https://docs.microsoft.com/en-us/dotnet/api/system.timespan.multiply).